### PR TITLE
Add support for v13 + General improvments

### PR DESCRIPTION
**Pull request by ParthDesai**
- Adding support for metadata v13
- Adding tests for metadata v13
- Modifying `CreateStorageKey` to be variadic for nmap
- Adding `FindConstantValue` for finding constant value from Metadata

Fixes: #159 

I have tested this on latest polkadot and it seems to parse the metadata perfectly.

### CODE REVIEW

**Review by vedhavyas — COMMENTED** [review]
Overall looks good. Will take another look again tomorrow.

Thanks you!

**Comment by ParthDesai** [comment]
There isn't any way we can remove the duplicate code linter error. Since it is required that the encoding/decoding happen in exact same order.

**Comment by ParthDesai** [comment]
@vedhavyas If everything looks alright to you, Can you merge it?

File: types/metadataV12.go [hunk 1]
```diff
@@ -83,6 +83,19 @@ func (m *MetadataV12) FindStorageEntryMetadata(module string, fn string) (Storag
 	return nil, fmt.Errorf("module %v not found in metadata", module)
 }
 
+func (m *MetadataV12) FindConstantValue(module Text, constant Text) ([]byte, error) {
```
Comment (vedhavyas) [inline types/metadataV12.go:86]: This looks like the repeat code. Why can't we reuse the one from Metadata V10

File: types/metadataV13.go [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+		}
+		if int(eventID[1]) >= len(mod.Events) {
+			return "", "", fmt.Errorf("event index %v for module %v out of range", eventID[1], mod.Name)
+		}
+		return mod.Name, mod.Events[eventID[1]].Name, nil
+	}
+	return "", "", fmt.Errorf("module index %v out of range", eventID[0])
+}
+
+func (m *MetadataV13) FindStorageEntryMetadata(module string, fn string) (StorageEntryMetadata, error) {
```
Comment (vedhavyas) [inline types/metadataV13.go:68]: same as above. Why not extend 13 from 12 and reuse those methods ?

**Review by vedhavyas — CHANGES_REQUESTED** [review]
Some of the new methods can be reused from extending the previous metadata version. That would also solve the linting issues I believe

File: types/metadataV13.go (hunk 2 above)
Comment (ParthDesai) [reply types/metadataV13.go:68]: @vedhavyas V13 is not just addition to V12, the storagemetadata enum changes.

Comment (vedhavyas) [reply types/metadataV13.go:68]: cool.

**Comment by vedhavyas** [comment]
@ParthDesai for linting issues you can add the `dupl` for golangci for those v12 files so that lint passes.

Tests aren't yet passing. Can you fix those pls

**Comment by ParthDesai** [comment]
@vedhavyas Tests are passing and solved linting issue by deriving `ModuleMetadataV12` from `ModuleMetadataV10`.

File: types/metadataV12.go (hunk 1 above)
Comment (ParthDesai) [reply types/metadataV12.go:86]: done

**Comment by ParthDesai** [comment]
> @ParthDesai for linting issues you can add the `dupl` for golangci for those v12 files so that lint passes.
> 
> Tests aren't yet passing. Can you fix those pls

@vedhavyas Tests are passing and solved linting issue by deriving `ModuleMetadataV12` from `ModuleMetadataV10`.  Please approve workflow run.

**Comment by canhlinh** [comment]
@ParthDesai waiting for these great works to be merged.

**Review by vedhavyas — APPROVED** [review]
Looks good!

Thank you

**Comment by ParthDesai** [comment]
@vedhavyas Is it possible to release a new version tag with v13 support? So, that dependent libraries can be updated.

**Comment by canhlinh** [comment]
@ParthDesai I just update manually by using the latest commit :D

**Comment by ParthDesai** [comment]
> @ParthDesai I just updated lib manually by using the latest commit :D

Yes, we can do that. But, it's better to point to a version tag than to point to a commit hash.

**Pull request merged**